### PR TITLE
Add POST Request

### PR DESCRIPTION
**Pull request by crudnicky**
# Motivation 

Michelle Slind has 45 unique targets that need to be uploaded to the PTDB. Rather than use the GUI to upload these targets, she requested that an uploader tool be created. This is the tool for the job.

## Background

Opening ./assets/XOLO_PTDB_Nomenclature.xlsx will show a table that contains the core data that is used to create new targets. Each row signifies a target that will be uploaded to the data, and the columns pertain to fields in a JSON object that is submitted to the API.

```
{
  "target": "",
  "partner": "",
  "protein_class_pk": "",
  "notes": "",
  "project": "",
  "subunits": [{
    "subunit_name": "",
    "copies": "1",
    "amino_acid_fasta_description": "V",
    "amino_acid_sequence": "",
    "genes": [{
      "dna_fasta_description": "",
      "dna_sequence": ""
    }]
  }, {
    "subunit_name": "",
    "copies": "1",
    "amino_acid_fasta_description": "",
    "amino_acid_sequence": "",
    "genes": [{
      "dna_fasta_description": "",
      "dna_sequence": ""
    }]
  }]
}
```

- `target` = Target Name
- `notes` = Description
-  `subunits[0] .name` = Subunit A name
-  `subunits[0].amino_acid_fasta_description && subunits[0].dna_fasta_description` = subunit a name
-  `subunits[1] .name` = Subunit B name
-  `subunits[1].amino_acid_fasta_description && subunits[0].dna_fasta_description` = subunit b name

Once this data is captured from the excel file the program will look in assets/CD19_MicAbody_Xolo_FASTA_files/ for the correct fasta file for the subunit and collect the corresponding sequence for each subunit.

After each target is created the program will pass the target to a method that uses Python's `requests` module to create a POST request to the correct endpoint.

___

## Notes

1. Michelle is going to likely be updating the excel file and the FASTA files so there might be some future refactors involving the logic behind creating names.
2. The POST request will not work without a TOKEN. The TOKEN value can be collected by following these steps: https://github.com/AbSciBio/protein_expression_front_end/blob/master/Authorization.pdf

### CODE REVIEW

**Review by rpfingsten — APPROVED** [review]
Looks good

File: main.py [hunk 1]
```diff
@@ -1,19 +1,60 @@
...
+    "Subunit 2 Fasta File IDs", 
+    "Subunit 2 AA sequence", 
+    "Subunit 2 DNA Sequence" 
+]
+for i in range(0, len(results_worksheet_column_names)):
+    results_worksheet.cell(1, i + 1, results_worksheet_column_names[i])
+targets = []
+
+# Upload data to PTDB
+for row in range(2, max_row + 1):
```
Comment (crudnicky) [inline main.py]: Make into a class

File: main.py [hunk 2]
```diff
@@ -1,19 +1,60 @@
...
+    results_worksheet.cell(i + 2, 4, targets[i]['project_name'])
+    results_worksheet.cell(i + 2, 5, targets[i]['subunits'][0]['subunit_name'])
+    results_worksheet.cell(i + 2, 6, targets[i]['subunits'][0]['subunit_name'])
+    results_worksheet.cell(i + 2, 7, targets[i]['subunits'][0]['amino_acid_sequence'])
+    results_worksheet.cell(i + 2, 8, targets[i]['subunits'][0]['genes'][0]['dna_sequence'])
+    results_worksheet.cell(i + 2, 9, targets[i]['subunits'][1]['subunit_name'])
+    results_worksheet.cell(i + 2, 10, targets[i]['subunits'][1]['subunit_name'])
+    results_worksheet.cell(i + 2, 11, targets[i]['subunits'][1]['amino_acid_sequence'])
+    results_worksheet.cell(i + 2, 12, targets[i]['subunits'][1]['genes'][0]['dna_sequence'])
+results_workbook.save('test.xlsx')
```
Comment (crudnicky) [inline main.py]: Make into class

Fix white space

File: parser.py [hunk 3]
```diff
@@ -3,55 +3,34 @@
...
-        subunit['genes'] = [(gene)]
-        return subunit
+        gene['dna_fasta_description'] = file_name
+        gene['dna_sequence'] = self.getFastaFileSequence(file_name, "DNA")
+        return [(gene)]
 
     def getFastaFileSequence(self, subunit, seq_type):
         file = file_reader.FileReader(subunit)
-        return file.get_fasta_description(seq_type)
+        return file.get_fasta_sequence(seq_type)
```
Comment (crudnicky) [inline parser.py]: remove whitespace

File: ptdb_request.py [hunk 4]
```diff
@@ -0,0 +1,22 @@
...
+    headers = self.get_headers()
+    json_target_data = self.target
+    # request = requests.post(URL, data=json_target_data, headers=headers)
+
+  def get_headers(self):
+    return {
+      "Authorization": f"Token {TOKEN}",
+      "Accept": "*/*",
+      "Content-Type": "application/json"
+    }
```
Comment (crudnicky) [inline ptdb_request.py]: add white space

**Review by crudnicky — COMMENTED** [review]
Refactor code a such that small class files are combined into one file with multiple classes and a tests directory is established. 

Add arg parser so that you can pass in a path to excel files, fasta files, and token.

Create an entry point method that can be called from the cli.

if __name__ == __target_uploader__

**Comment by jeads-absci** [comment]
To clarify, the way to make the entry point in the main CLI script is to add the following conditional block

```
def entry_point_function():
    # Run main execution flow here

if __name__ == "__main__":
    entry_point_function()
```

**Pull request merged**